### DISCUSSION TITLE
Configure case-sensitivity in sorting

### DISCUSSION THREAD

**Discussion by damnsamn** [payloadcms/payload#8005] [Feature Requests & Ideas]
Hey guys - we’ve encountered an issue with our site where sorting a collection by its `name` field via the REST API (`GET /api/posts?limit=0&sort=name`) is returning alphabetical sorting in unexpected ways, seeming to sort Capitals-first, where `AZ` precedes `Aa`, and `Z` precedes `a`.

Below are two screenshots where this is an issue:
![IMG_3428](https://github.com/user-attachments/assets/9dc2399a-b2b5-4acc-ae70-4469dd048b86)
![IMG_3429](https://github.com/user-attachments/assets/699beff7-959c-4d2b-8f83-4e171fc89aaf)

1. `NSO Group` sorted above `Narus` - we’d expect it to be sorted alphabetically under `Nodus Information Technology`
2. Stylised names that start with lowercase such as `iFLYTEK` are sorted to the absolute bottom of the list, underneath all items written as proper nouns

It would be great if there were a way to configure payload’s sorting algorithm to sort in a way that is case-insensitive, or at least sorting alphabetical->casing rather than casing->alphabetical.

**Comment by hugoburguete** [comment]
I know this is half a year old but did you find an answer in the end? I am facing the same issue and it would be nice if we could have a case insensitive sort

**Reply by damnsamn** [reply]
No, we ended up just moving the sorting logic to the front-end. Maybe if you were to move this discussion into an actual issue it might receive some stronger attention?

**Reply by damnsamn** [reply]
> I tested a few approaches and I think the ideal one until we figure out this feature for the core APIs would be to make an admin-hidden field such as sortValue that you can use for sorting, using a beforeChange hook you can populate it with the title lowercased with toLowerCase()

This was the response from one of the payload team members in a private slack channel though - this is probably a good interim solution

**Reply by hugoburguete** [reply]
Thanks.

Luckily I can make sure that when my collection value is saved, I am able to capitalize the name, so that will be my workaround. It would be good if we could create a custom sort (that doesn't involve adding a dummy field).

**Comment by paulpopus** [comment]
[Issue #14517](https://github.com/payloadcms/payload/issues/14517) describes the same underlying request, specifically for the system-managed `filename` field in Media. I’m closing that issue in favor of keeping the broader request and implementation discussion here.

In v4, in [#17741](https://github.com/payloadcms/payload/pull/17741) I added `query.operatorHandlers`. That provides a useful architectural pattern, but it cannot solve case-insensitive sorting directly because those handlers affect `WHERE` comparisons, while sorting goes through a separate `ORDER BY` path.

One possible v4 approach would be a sibling adapter-level extension point such as `query.sortHandlers` or `query.orderByHandlers`. After resolving the requested sort field, Payload could pass the adapter, collection or table, field path, column, direction, and locale to matching handlers. A handler could then transform the ordering expression—for example using `LOWER(column)` or an adapter-specific collation—before falling back to the default sort behavior.

Ideally, we would provide a built-in case-insensitive sorting preset or higher-level configuration, while retaining sort handlers as the lower-level escape hatch.
